storage: Add range key buffering to MultiSSTWriter

raggar opened the pull request:
This pr consists of two main changes. The first is moving the `MultiSSTWriter` to package `storage`. The second change comes from the following motivation.

Currently, when receiving a snapshot, we ingest a fixed number of sstables, corresponding to the range's various contiguous keyspaces. When the default range size increased from 64 MB to 512 MB, we started ingesting user data sstables up to 512 MB. These large files cause more expensive compactions. As a result of limiting the size of SSTables, this pr adds range key buffering and truncation to the existing MultiSSTWriter logic.

NEED TO FIX: 
- Currently TestMultiSSTWriterBuffer test is failing (needs to be fixed --> code seems to be adding to the internal SSTWriter fragmenter that claims to be closed)

TODO:
- Run tests (can't do it currently because fragmenter is not exported)
- Replace backups case to use the MultiSSTWriter (can maybe do this in a separate pr) 

Release note: None
Informs: #67284

Commit:
storage: Add range key buffering to MultiSSTWriter

Currently, when receiving a snapshot, we ingest a
fixed number of sstables, corresponding to the range's various
contiguous keyspaces. When the default range size increased from 64 MB
to 512 MB, we started ingesting user data sstables up to 512 MB. These
large files cause more expensive compactions. As a result of limiting
the size of SSTables, this pr adds range key buffering and truncation to
the existing MultiSSTWriter logic.

Release note: None
Informs: #67284

jbowens reviewed (commented):
Can you please separate the two changes (the move to pkg/storage and then the range key buffering) into two separate commits? It makes it easier to review how the MultiSSTWriter has changed to accommodate range keys.

You may already know of a way to separate the commits, but if not the way I would do it is: Check out the commit's parent in a new branch (`git checkout sst_writer_with_buffer_truncation^ -b move-sst-writer`), perform just the move of the `MultiSSTWriter` to pkg/storage and commit it. Then back on this branch you can `git rebase -i HEAD^^` and `pick` the commit between the parent commit and this existing commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-:-N_FxLWh2jBOgRgaNqjh:b-28brpz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RahulAggarwal1016)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_FwjHG7XRf7ZnnGPoY:-N_FwjHG7XRf7ZnnGPoZ:bvu0956) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f7123f5756167dcab40f0796f9fabf60dd5231c/pkg/kv/kvserver/store_snapshot.go#L322)):*
> ```Go
> 		writer := storage.MakeIngestionSSTWriter(ctx, kvSS.st, newSSTFile)
> 		return writer, nil
> 	}
> ```

I think we should be able to define this func (`createNewSSTWriter`) as a method on `kvBatchSnapshotStrategy`, and we might be able to prevent the code duplication in `TestMultiSSTWriterInitSST` that way.


<!-- Sent from Reviewable.io -->

Commit:
kvserver: Move MultiSSTWriter to storage package

This pr moves MultiSSTWriter to the storage package. This is done so
that its logic can be reused in multiple different places such as
backups.

Release note: None
Informs: #67824

raggar reviewed (commented):
done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-:-N_GBMo-E8hmH8LIFbXZ:bb7f2mf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_FwjHG7XRf7ZnnGPoY:-N_GBOlY3oBzxhilj1Mz:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f7123f5756167dcab40f0796f9fabf60dd5231c/pkg/kv/kvserver/store_snapshot.go#L322)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we should be able to define this func (`createNewSSTWriter`) as a method on `kvBatchSnapshotStrategy`, and we might be able to prevent the code duplication in `TestMultiSSTWriterInitSST` that way.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
> Reviewed 1 of 4 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RahulAggarwal1016)

___
*[`pkg/storage/multi_sst_writer.go` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_isZnl6M3SRUSWsMlY:-N_isZnl6M3SRUSWsMlZ:btx3esb) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L11)):*
> ```Go
> // the Business Source License, use of this software will be governed
> // by the Apache License, Version 2.0, included in the file
> // licenses/APL.txt.
> ```

The copyright notice should appear first in the file, then a blank line, then the package declaration.
___
*[`pkg/storage/multi_sst_writer.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_isxjw604NgUvFouPv:-N_isxjw604NgUvFouPw:bl4l95o) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L21)):*
> ```Go
> 
> // multiSSTWriter is a wrapper around an SSTWriter and SSTSnapshotStorageScratch
> // that handles chunking SSTs and persisting them to disk.
> ```

nit: This comment is out-of-date. How about something like:

```
// MultiSSTWriter may be used to construct a series of non-overlapping sstables
// from a stream of keys. The user may specify a set of fixed boundaries such
// that no sstable produced will span the boundaries.
```
___
*[`pkg/storage/multi_sst_writer.go` line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_ismW-7JOjffEWU_Iu:-N_ismW-7JOjffEWU_Iv:baxf1t5) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L22)):*
> ```Go
> // multiSSTWriter is a wrapper around an SSTWriter and SSTSnapshotStorageScratch
> // that handles chunking SSTs and persisting them to disk.
> type multiSSTWriter struct {
> ```

We should export this type now that it's in the storage package and will be consumed by other packages.
___
*[`pkg/storage/multi_sst_writer.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_itfkm75zXciRRgQDv:-N_itfkm75zXciRRgQDw:b-vssarq) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L55)):*
> ```Go
> 	if err := msstw.currSST.ClearRawRange(
> 		msstw.keySpans[msstw.currSpan].Key, msstw.keySpans[msstw.currSpan].EndKey,
> 		true /* pointKeys */, true, /* rangeKeys */
> ```

This call to ClearRawRange only applies to the snapshot reception case (eg, will not apply to backups). Additionally this will also need to be buffered and split at split points.
___
*[`pkg/storage/multi_sst_writer.go` line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_iuVdD8W3mG-BUALz3:-N_iuVdD8W3mG-BUALz4:b3by2x0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8c56b438bcd0225eb7d2ff6461a70ef17571120/pkg/storage/multi_sst_writer.go#L15)):*
> ```Go
> )
> 
> const SSTSizeLimit = 128 << 20
> ```

nit: Let's make this a parameter on `NewMultiSSTWriter`.


<!-- Sent from Reviewable.io -->

raggar reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/storage/multi_sst_writer.go` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_isZnl6M3SRUSWsMlY:-N_jTZR37kOR2fneKQLE:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L11)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

The copyright notice should appear first in the file, then a blank line, then the package declaration.
</blockquote></details>

done
___
*[`pkg/storage/multi_sst_writer.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_isxjw604NgUvFouPv:-N_jUVqF6hXN0lrgOqGw:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L21)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: This comment is out-of-date. How about something like:

```
// MultiSSTWriter may be used to construct a series of non-overlapping sstables
// from a stream of keys. The user may specify a set of fixed boundaries such
// that no sstable produced will span the boundaries.
```
</blockquote></details>

done
___
*[`pkg/storage/multi_sst_writer.go` line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_ismW-7JOjffEWU_Iu:-N_jf23R8kLBB1zN-JuS:b-7rl0jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L22)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

We should export this type now that it's in the storage package and will be consumed by other packages.
</blockquote></details>

done 
___
*[`pkg/storage/multi_sst_writer.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_itfkm75zXciRRgQDv:-N_jf41V9Q2nboHeiqEM:bq80a1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L55)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

This call to ClearRawRange only applies to the snapshot reception case (eg, will not apply to backups). Additionally this will also need to be buffered and split at split points.
</blockquote></details>

when writing the `rangeDel` through ClearRawRange, does the rangeDel's (start and end) have to be in order with respect to the other point keys? `SSTWriter.ClearRawRange` for the `multiSSTWriter` calls both `DeleteRange` and RangeKeyDelete` 
___
*[`pkg/storage/multi_sst_writer.go` line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_iuVdD8W3mG-BUALz3:-N_jUBVWAEMWLwkvC37o:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8c56b438bcd0225eb7d2ff6461a70ef17571120/pkg/storage/multi_sst_writer.go#L15)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: Let's make this a parameter on `NewMultiSSTWriter`.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RahulAggarwal1016)

___
*[`pkg/storage/multi_sst_writer.go` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-N_itfkm75zXciRRgQDv:-Na8Xzd57iOtKZ6kYkp6:bc707v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/152678d80652512e09e2e19928a633eee12a97db/pkg/storage/multi_sst_writer.go#L55)):*
> does the rangeDel's (start and end) have to be in order with respect to the other point keys?

It does not. Range dels need to be in order with respect to other range dels, and range keys need to be in order with respect to other range keys.


<!-- Sent from Reviewable.io -->

RahulAggarwal1016 commented on pkg/storage/multi_sst_writer.go:
```diff
@@ -0,0 +1,222 @@
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/pebble/internal/base"
+	"github.com/cockroachdb/pebble/internal/keyspan"
```
Need to wait for https://github.com/cockroachdb/pebble/pull/2815 to be merged

RahulAggarwal1016 commented on pkg/storage/multi_sst_writer.go:
```diff
@@ -0,0 +1,250 @@
+
+// MultiSSTWriter may be used to construct a series of non-overlapping sstables
+// from a stream of keys. The user may specify a set of fixed boundaries such
+// that no sstable produced will span the boundaries.
```
This comment was here previously not sure why it says "no sstable produced will span the boundaries", don't we want the opposite (i.e. to have SST's that span the `keyspans`)?

RahulAggarwal1016 commented on pkg/storage/multi_sst_writer.go:
```diff
@@ -0,0 +1,250 @@
+	// Used for fragmenting range keys.
+	rangeKeyFrag keyspan.Fragmenter
+	// Used for fragmenting range deletes.
+	rangeDelFrag keyspan.Fragmenter
```
`sstable.Writer` also has a `fragmenter` (maybe it serves a similar purpose) but I was not too sure

jbowens reviewed (commented):
Looking really good!

> Reviewed 5 of 5 files at r7, 3 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-:-NbuwtUV-mjCuzo4bUYO:b-b00xdb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RahulAggarwal1016 and @sumeerbhola)

___
*[`pkg/storage/multi_sst_writer.go` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-Nbuf9yPD_S-EpuZUwTi:-Nbuf9yPD_S-EpuZUwTj:b-cr2q3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f7123f5756167dcab40f0796f9fabf60dd5231c/pkg/storage/multi_sst_writer.go#L9)):*
> ```Go
> // the Business Source License, use of this software will be governed
> // by the Apache License, Version 2.0, included in the file
> // licenses/APL.txt.
> ```

This copyright notice should be the Cockroach copyright notice, not the Pebble notice. The two repositories have separate licenses.
___
*[`pkg/storage/multi_sst_writer.go` line 21 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/106771#gh-1291457070:-NbunLiO3eH0UbMz7WcX:buoxdpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f2b4bd9f8fa310d2638f9b57f6a9f33385c9198/pkg/storage/multi_sst_writer.go#L21)):*
<details><summary><i>Previously, RahulAggarwal1016 (Rahul Aggarwal) wrote…</i></summary><blockquote>

This comment was here previously not sure why it says "no sstable produced will span the boundaries", don't we want the opposite (i.e. to have SST's that span the `keyspans`)? 
</blockquote></details>

I think this is saying that none of the sstables produced will overlap one of the boundaries. The boundaries are fragmentation points, preventing any single sstable from containing keys on both sides of the boundary.
___
*[`pkg/storage/multi_sst_writer.go` line 162 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/106771#gh-1293656625:-NbuwZ2J8n6qJ5kXEyf_:bju45b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1f32c2f046f90991981b2f7ad675c2feaad44d6/pkg/storage/multi_sst_writer.go#L162)):*
<details><summary><i>Previously, RahulAggarwal1016 (Rahul Aggarwal) wrote…</i></summary><blockquote>

we probably want to include the size of the currently buffered range keys to have our file size be more strict? If so can maybe copy something like this https://github.com/cockroachdb/cockroach/blob/2b18f4ffc260dca938ba0265217c4483ec6064c7/pkg/storage/mvcc.go#L6632 

We might need to modify the `fragmenter` to access the number of pending/buffered keys
</blockquote></details>

Let's leave a TODO noting that it currently doesn't include buffered range keys and range deletions. I think this is fine. Pebble itself also ignores range deletions and range keys sizes during compactions. They're expected to be infrequent and inconsequential.
___
*[`pkg/storage/multi_sst_writer.go` line 26 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-Nbuw8sF2LqZQ_QMsRMR:-Nbuw8sF2LqZQ_QMsRMS:b-jmepyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/storage/multi_sst_writer.go#L26)):*
> ```Go
> 	getNewSSTFile func(ctx context.Context) (SSTWriter, error)
> 	// uponFinish is called once an
> 	beforeSSTFinalize func(start, end roachpb.Key) error
> ```

unused
___
*[`pkg/storage/multi_sst_writer.go` line 59 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-NburuLa0fQV7rMH0akK:-NburuLa0fQV7rMH0akL:b-k3rgz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/storage/multi_sst_writer.go#L59)):*
> ```Go
> 			for _, key := range span.Keys {
> 				if err := msstw.currSST.PutEngineRangeKey(span.Start, span.End, key.Suffix, key.Value); err != nil {
> 					log.Warningf(ctx, "%v", err)
> ```

Can we add an `err` field to `MultiSSTWriter`,  record this error to the field and surface it on the start of the next fallible method call? We'll also want to record the error in the rangedel case to `err`.
___
*[`pkg/storage/multi_sst_writer.go` line 115 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-Nbut2mi7LEBAB4tLQCR:-Nbut2mi7LEBAB4tLQCS:brtc8ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/storage/multi_sst_writer.go#L115)):*
> ```Go
> 			{Trailer: uint64(pebble.InternalKeyKindDelete)},
> 		},
> 	})
> ```

The details of clearing these spans on init belong in `kvserver/store_snapshot.go`. One idea to achieve that:

1. `NewMultiSSTWriter` takes a `onEnterSpan func(context.Context, roachpb.Span, *MultiSSTWriter) error` 
2. `kvserver/store_snapshot.go` passes in an argument:
```
func clearSpan(ctx context.Context, span roachpb.Span, w *MultiSSTWriter) error {
    // Clear all existing replica state, including both point keys and range keys.
    if err := w.PutRangeDel(ctx, span.Key, span.EndKey); err != nil {
        return err    
    }
    if err := w.PutRangeKeyDel(ctx, span.Key, span.EndKey); err != nil {
        return err
    }
    return nil
}
```

Then this `clearCurrSpan` function becomes `initializeNewSpan` and calls `w.clearSpan`.
___
*[`pkg/storage/multi_sst_writer.go` line 153 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-NburU4B1z5ycf7vkjXC:-NburU4B1z5ycf7vkjXD:b4w7brb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/storage/multi_sst_writer.go#L153)):*
> ```Go
> 			return err
> 		}
> 		// Since our span has been incremented for the first time, we must clear the current span.
> ```

nit: how about "Since we've moved on to a new span, we must initialize the new span."
___
*[`pkg/storage/multi_sst_writer.go` line 214 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-Nbuvjpu53syx8UaRVBp:-Nbuvjpu53syx8UaRVBq:b-g508a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/storage/multi_sst_writer.go#L214)):*
> ```Go
> }
> 
> func (msstw *MultiSSTWriter) PutRangeDel(ctx context.Context, start, end roachpb.Key) error {
> ```

We should also expose a method of writing RangeKeyDels (see my other comment on pushing the logic that writes RangeKeyDels and RangeDels into kvserver.)
___
*[`pkg/kv/kvserver/testdata/multi_sst_writer` line 12 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-NbuqE639Y9_AFR1Aopp:-NbuqE639Y9_AFR1Aopq:b-fz4sm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/kv/kvserver/testdata/multi_sst_writer#L12)):*
> ```
> ----
> Table 1
> Point keys: 0
> ```

Can we include the tables' bounds here? Eg, `metadata.SmallestPointKey`,`metadata.LargestPointKey` and `metadata.SmallestRangeKey`,`mretadata.LargestRangeKey`
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 310 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/106771#-NbuqyOHAc7g_aXn4htl:-NbuqyOHAc7g_aXn4htm:b31eecd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f29b8ed89f29d08ea5f4b08052b83a189dcc3304/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L310)):*
> ```Go
> 			return b.String()
> 		case "write":
> 			//var b strings.Builder
> ```

nit: detritus?


<!-- Sent from Reviewable.io -->
